assembly/src/assembler/module_graph/mod.rs

```diff
@@ -238,9 +241,27 @@ impl ModuleGraph {
 // ------------------------------------------------------------------------------------------------
 /// Kernels
 impl ModuleGraph {
-    pub(super) fn set_kernel(&mut self, kernel_index: Option<ModuleIndex>, kernel: Kernel) {
-        self.kernel_index = kernel_index;
-        self.kernel = kernel;
+    pub(super) fn with_kernel(kernel_module: ModuleInfo) -> Result<Self, AssemblerError> {
+        if kernel_module.path() != &LibraryPath::from(LibraryNamespace::Kernel) {
+            return Err(AssemblerError::NoKernelModuleInKernelLibrary);
+        }
+
+        // build the kernel struct from the kernel module
+        let proc_hashes: Vec<_> = kernel_module.procedure_digests().collect();
+        let kernel = Kernel::new(&proc_hashes)?;
+
+        // add kernel module to the graph
+        // TODO: simplify this to avoid using Self::add_compiled_modules()
```

Comment:
Should we remove this TODO and create an issue for this? Or did you intend to do it in this PR?

Reply:
This is intended for a follow-up PR - but probably not worth creating an issue for just this. I think this will be covered in #1416.